### PR TITLE
TEAMU-940 Add new migration 41 replacing MAX and MIN with last and first

### PR DESCRIPTION
**Pull request by wtcarter**
In an attempt to fix anomolies in Hilltop data views, replacing MIN(reading) and MAX(reading) to first and last non-zero values for observed_water_use_aggregated_daily materialized view

### CODE REVIEW

File: packages/Manager/src/main/resources/db/migration/V0041__observed_water_use_aggregated_daily.sql [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+         WHERE amount != 0
+     ),
+     daily_calculation AS (
+SELECT filtered_obs.site_name,
+       filtered_obs.day_observed_at,
+       filtered_obs.measurement_name,
+       CASE
+           WHEN filtered_obs.measurement_name = 'Water Meter Volume' THEN SUM(filtered_obs.amount)
+           WHEN count(*) = 1 AND LAG(max(filtered_obs.day_observed_at)) OVER (PARTITION BY id, filtered_obs.measurement_name ORDER BY id, filtered_obs.day_observed_at, filtered_obs.measurement_name) != filtered_obs.day_observed_at - interval '1 day' THEN NULL
+           WHEN count(*) = 1 THEN first_values.first_value - last_values.last_value
```
Comment (MartyP233) [inline packages/Manager/src/main/resources/db/migration/V0041__observed_water_use_aggregated_daily.sql:41]: this row (41) should remain what it was - the WHEN conditions get met when there is 1 reading per day, so it needs to take the 1 reading from that day and subtract the previous day to get usage for that day.

File: packages/Manager/src/main/resources/db/migration/V0041__observed_water_use_aggregated_daily.sql [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+           WHEN filtered_obs.measurement_name = 'Water Meter Reading' THEN first_values.first_value - last_values.last_value
```
Comment (MartyP233) [inline packages/Manager/src/main/resources/db/migration/V0041__observed_water_use_aggregated_daily.sql:42]: would it work to aggregate here: 
`WHEN filtered_obs.measurement_name = 'Water Meter Reading' THEN LAST(filtered_obs.amount) - FIRST(filtered_obs.amount)`
and remove the need for the first_values and last_values aggregation step?

**Review by MartyP233 — COMMENTED** [review]
hey Wayne - I've dropped a couple of comments, let me know what you think or if i'm barking up the wrong tree.

**Comment by wtcarter** [comment]
Yep, seen them.  They look good and def help understand things a bit better to write the tests.  I’ll wait till I’m settled on the train to check them out.  We’re just about to load up :)

Wayne
________________________________

**Comment by MartyP233** [comment]
also noticed wayne that this migration has the sql for 2 views in it. you're just changing observed_water_use_aggregated_daily so you can remove the materialised view logic.

**Comment by wtcarter** [comment]
That was my first version actually.  I just added the CTEs (broke it down) while I was analysing the bits of the query, but I do agree they’re not really necessary.  I’ll remove them

Wayne
________________________________

File: packages/Manager/src/main/resources/db/migration/V0041__observed_water_use_aggregated_daily.sql (hunk 2 above)
Comment (wtcarter) [reply packages/Manager/src/main/resources/db/migration/V0041__observed_water_use_aggregated_daily.sql:42]: As far as I can tell, Postgres has no LAST function, only the more complex LAST_VALUE

**Pull request merged**